overlord/ifacestate/ifacestate.go

```diff
@@ -35,17 +36,90 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
-var noConflictOnConnectTasks = func(task *state.Task) bool {
-	return task.Kind() != "connect" && task.Kind() != "disconnect"
+func noConflictOnConnectTasks(chg *state.Change) func(task *state.Task) bool {
+	checkConflict := func(task *state.Task) bool {
+		if chg != nil && task.Change() != nil {
+			// if same change, then return false (no conflict)
+			return chg.ID() != task.Change().ID()
+		}
+		return task.Kind() != "connect" && task.Kind() != "disconnect"
+	}
+	return checkConflict
+}
+
+var connectRetryTimeout = time.Second * 5
+
+func checkConnectConflicts(st *state.State, change *state.Change, plugSnap, slotSnap string) error {
+	for _, chg := range st.Changes() {
+		if chg.Status().Ready() || (change != nil && change.ID() == chg.ID()) {
+			continue
+		}
+		if chg.Kind() == "transition-ubuntu-core" {
+			return fmt.Errorf("ubuntu-core to core transition in progress, no other changes allowed until this is done")
+		}
+	}
+
+	for _, task := range st.Tasks() {
+		k := task.Kind()
+		chg := task.Change()
+
+		// there is no conflict if same change
+		if change != nil && chg != nil && change.ID() == chg.ID() {
```

Comment:
I don't think this is exactly what we want here, I think in the case of auto-connect we need to get the name of the snap being installed and ignore it. OTOH given that we wait from the auto-connect task on link-snap/setup-profiles tasks, there's no problem waiting even in the same change (as long as I said we ignore the snap being installed)

Reply:
Good point, thanks. Done.